src/ts/forwarder.ts

```diff
@@ -0,0 +1,188 @@
+import { ChainID } from "..";
+import { ReplayProtectionAuthority } from "./replayprotectionauthority";
+import { Contract, Wallet } from "ethers";
+import {
+  defaultAbiCoder,
+  BigNumberish,
+  arrayify,
+  keccak256,
+} from "ethers/utils";
+
+export interface ForwardParams {
+  to: string;
+  signer: string;
+  target: string;
+  value: string;
+  data: string;
+  replayProtection: string;
+  replayProtectionAuthority: string;
+  chainId: number;
+  signature: string;
+}
+
+export interface DeploymentParams {
+  to: string;
+  signer: string;
+  initCode: string;
+  replayProtection: string;
+  replayProtectionAuthority: string;
+  chainId: number;
+  signature: string;
+}
+
+export interface RelayCallData {
+  target: string;
+  callData: string;
+}
+
+export interface ProxyCallData {
+  target: string;
+  value: BigNumberish;
+  callData: string;
+}
+
+/**
+ * Provides common functionality for the RelayHub and the ProxyAccounts.
+ * Possible to extend it with additional functionality if another
+ * msg.sender solution emerges.
+ */
+export abstract class Forwarder<T> {
+  constructor(
+    protected readonly chainID: ChainID,
+    protected readonly forwarder: Contract,
```

Comment:
`this.forwarder` is never referenced in this base class, you should remove it from here. Also, you're incorrectly populating it with the account factory in some places.

Reply:
sorted